File: src/composables/handlers/filesystem.ts
import { save } from '@tauri-apps/plugin-dialog'
import { writeFile } from '@tauri-apps/plugin-fs'
import { openPath } from '@tauri-apps/plugin-opener'
import { tempDir, join } from '@tauri-apps/api/path'
import { HAEXTENSION_METHODS } from '@haexhub/sdk'
import type { IHaexHubExtension } from '~/types/haexhub'
import type { ExtensionRequest } from './types'

export async function handleFilesystemMethodAsync(
  request: ExtensionRequest,
  extension: IHaexHubExtension,
) {
  if (!request || !extension) return

  switch (request.method) {
    case HAEXTENSION_METHODS.filesystem.saveFile: {
      const params = request.params as {
        data: number[]
        defaultPath?: string
        title?: string
        filters?: Array<{ name: string; extensions: string[] }>
      }

      // Convert number array back to Uint8Array
      const data = new Uint8Array(params.data)

      // Open save dialog
      const filePath = await save({
        defaultPath: params.defaultPath,
        title: params.title || 'Save File',
        filters: params.filters,
      })

      // User cancelled
      if (!filePath) {
        return null
      }

      // Write file
      await writeFile(filePath, data)

      return {
        path: filePath,
        success: true,
      }
    }

    case HAEXTENSION_METHODS.filesystem.showImage: {
      // This method is now handled by the frontend using PhotoSwipe
      // We keep it for backwards compatibility but it's a no-op
      return {
        success: true,
        useFrontend: true,
      }
    }

    case HAEXTENSION_METHODS.filesystem.openFile: {
      const params = request.params as {
        data: number[]
        fileName: string
        mimeType?: string
      }

      try {
        // Convert number array back to Uint8Array
        const data = new Uint8Array(params.data)

        // Get temp directory and create file path
        const tempDirPath = await tempDir()
        const tempFilePath = await join(tempDirPath, params.fileName)

        // Write file to temp directory
        await writeFile(tempFilePath, data)

        // Open file with system's default viewer
        await openPath(tempFilePath)

        return {
          success: true,
        }
      }
      catch (error) {
        console.error('[Filesystem] Error opening file:', error)
        return {
          success: false,
        }
      }
    }

    default:
      throw new Error(`Unknown filesystem method: ${request.method}`)
  }
}

File: src-tauri/src/extension/webview/filesystem.rs
use crate::extension::error::ExtensionError;
use crate::AppState;
use serde::{Deserialize, Serialize};
use tauri::{State, WebviewWindow};
use tauri_plugin_dialog::DialogExt;
use tauri_plugin_opener::OpenerExt;

#[derive(Debug, Clone, Deserialize)]
pub struct FileFilter {
    pub name: String,
    pub extensions: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SaveFileResult {
    pub path: String,
    pub success: bool,
}

#[tauri::command]
pub async fn webview_extension_fs_save_file(
    window: WebviewWindow,
    _state: State<'_, AppState>,
    data: Vec<u8>,
    default_path: Option<String>,
    title: Option<String>,
    filters: Option<Vec<FileFilter>>,
) -> Result<Option<SaveFileResult>, ExtensionError> {
    eprintln!("[Filesystem] save_file called with {} bytes", data.len());
    eprintln!("[Filesystem] save_file default_path: {:?}", default_path);
    eprintln!("[Filesystem] save_file first 10 bytes: {:?}", &data[..data.len().min(10)]);

    // Build save dialog
    let mut dialog = window.dialog().file();

    if let Some(path) = default_path {
        dialog = dialog.set_file_name(&path);
    }

    if let Some(t) = title {
        dialog = dialog.set_title(&t);
    }

    if let Some(f) = filters {
        for filter in f {
            let ext_refs: Vec<&str> = filter.extensions.iter().map(|s| s.as_str()).collect();
            dialog = dialog.add_filter(&filter.name, &ext_refs);
        }
    }

    // Show dialog (blocking_save_file is safe in async commands)
    eprintln!("[Filesystem] Showing save dialog...");
    let file_path = dialog.blocking_save_file();

    if let Some(file_path) = file_path {
        // Convert FilePath to PathBuf
        let path_buf = file_path.as_path().ok_or_else(|| ExtensionError::ValidationError {
            reason: "Failed to get file path".to_string(),
        })?;

        eprintln!("[Filesystem] User selected path: {}", path_buf.display());
        eprintln!("[Filesystem] Writing {} bytes to file...", data.len());

        // Write file using std::fs
        std::fs::write(path_buf, &data)
            .map_err(|e| {
                eprintln!("[Filesystem] ERROR writing file: {}", e);
                ExtensionError::ValidationError {
                    reason: format!("Failed to write file: {}", e),
                }
            })?;

        eprintln!("[Filesystem] File written successfully");

        Ok(Some(SaveFileResult {
            path: path_buf.to_string_lossy().to_string(),
            success: true,
        }))
    } else {
        eprintln!("[Filesystem] User cancelled");
        // User cancelled
        Ok(None)
    }
}

#[tauri::command]
pub async fn webview_extension_fs_open_file(
    window: WebviewWindow,
    _state: State<'_, AppState>,
    data: Vec<u8>,
    file_name: String,
) -> Result<serde_json::Value, ExtensionError> {
    // Get temp directory
    let temp_dir = std::env::temp_dir();
    let temp_file_path = temp_dir.join(&file_name);

    // Write file to temp directory using std::fs
    std::fs::write(&temp_file_path, data)
        .map_err(|e| ExtensionError::ValidationError {
            reason: format!("Failed to write temp file: {}", e),
        })?;

    // Open file with system's default viewer
    let path_str = temp_file_path.to_string_lossy().to_string();
    window.opener().open_path(path_str, None::<String>)
        .map_err(|e| ExtensionError::ValidationError {
            reason: format!("Failed to open file: {}", e),
        })?;

    Ok(serde_json::json!({
        "success": true
    }))
}
